controllers/metal3.io/host_state_machine.go

```diff
@@ -226,6 +230,34 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
+func (hsm *hostStateMachine) checkDeregisteredHost(info *reconcileInfo) (result actionResult) {
+	// If the deregistered annotation is set we remove any host from the
+	// provisioner and take no further action
+	// Note this doesn't change the current state, only the OperationalStatus
+	annotations := hsm.Host.GetAnnotations()
+	if annotations != nil {
+		if _, ok := annotations[metal3v1alpha1.DeregisteredAnnotation]; ok {
```

Comment:
Here would be the place to switch on the state if we decide to do that.
We're thinking we don't want this to happen in the [De]Provisioning states.
We also don't want to change the operational status in the Unmanaged state imho, although we do still want to return early.

It occurs to me that if we leave IPA running (i.e. in any state except [Externally]Provisioned) then the hosts will get rediscovered by ironic and the nodes recreated. I guess if we don't move them to the manageable state that's OK though?

Reply:
> It occurs to me that if we leave IPA running (i.e. in any state except [Externally]Provisioned) then the hosts will get rediscovered by ironic and the nodes recreated. I guess if we don't move them to the manageable state that's OK though?

That's assuming a discovery workflow exists, which it doesn't today, right?

I'd expect any discovery flow to be smart enough to avoid creating duplicate BMH resources when an existing one exists (it'd have to rely on matching `bootMACAddress` I guess) - that's a slightly different case than was solved by @honza in https://github.com/metal3-io/baremetal-operator/pull/780 though (which relies on matching the mac at the Ironic level AFAICS.

I guess one option is to only action this annotation when the state is `Provisioned` - then we know IPA is done, and the only valid transition from that point will be deleting (without deprovisioning), unless the annotation is removed?

Reply:
I think in practice this won't be a problem, the only use-case I'm aware of is to detach Provisioned BMH resources on the parent cluster, and as you say provided we don't move the hosts to a state where Ironic asserts power state things will work fine even if in a corner-case scenario IPA does run and discover a host.

If we need to enforce this, it may be better to only allow detaching in [Externally]Provisioned - another option now that the implementation uses unmanagedRetryDelay is we could call `prov.Detach()` unconditionally, so any hanging Ironic resources would be eventually removed?

Reply:
The latest version restricts the API to  [Externally]Provisioned)